pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix

```diff
@@ -0,0 +1,88 @@
+{ pkgsi686Linux
+, stdenv
+, fetchurl
+, dpkg
+, makeWrapper
+, coreutils
+, ghostscript
+, gnugrep
+, gnused
+, which
+, perl
+, lib
+}:
+
+let 
+  model = "mfcl3770cdw";
+  version = "1.0.2-0";
+  src = fetchurl {
+    url = "https://download.brother.com/welcome/dlf103935/${model}pdrv-${version}.i386.deb";
+    sha256 = "09fhbzhpjymhkwxqyxzv24b06ybmajr6872yp7pri39595mhrvay";
+  };
+  reldir = "opt/brother/Printers/${model}/";
+
+in rec { 
+  driver = stdenv.mkDerivation rec {
+    inherit src version;
+    name = "${model}drv-${version}";
+
+    nativeBuildInputs = [ dpkg makeWrapper ];
+
+    unpackPhase = "dpkg-deb -x $src $out";
+
+    installPhase = ''
+      dir="$out/${reldir}"
+      substituteInPlace $dir/lpd/filter_${model} \
+        --replace /usr/bin/perl ${perl}/bin/perl \
+        --replace "BR_PRT_PATH =~" "BR_PRT_PATH = \"$dir\"; #" \
+        --replace "PRINTER =~" "PRINTER = \"${model}\"; #"
+      wrapProgram $dir/lpd/filter_${model} \
+        --prefix PATH : ${stdenv.lib.makeBinPath [
+          coreutils ghostscript gnugrep gnused which
+        ]}
+    # need to use i686 glibc here, these are 32bit proprietary binaries
+    interpreter=${pkgsi686Linux.glibc}/lib/ld-linux.so.2
+    patchelf --set-interpreter "$interpreter" $dir/lpd/brmfcl3770cdwfilter
```

Comment:
I think using the `$(cat $NIX_CC/nix-support/dynamic-linker)`  way to figure out the right `ld-linux.so` is probably better. Since this is called using `pkgsi686Linux.callPackage` it'll end up relying on the right ld-linux interpreter (I think). A tad annoying that (in that particular driver) they only ship 32 bit binaries.

```diff
diff --git a/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix b/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
index 9fa1c052597..658fe3af77c 100644
--- a/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
+++ b/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
@@ -40,9 +40,8 @@ in rec {
         --prefix PATH : ${stdenv.lib.makeBinPath [
           coreutils ghostscript gnugrep gnused which
         ]}
-    # need to use i686 glibc here, these are 32bit proprietary binaries
-    interpreter=${pkgsi686Linux.glibc}/lib/ld-linux.so.2
-    patchelf --set-interpreter "$interpreter" $dir/lpd/brmfcl3770cdwfilter
+      patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+        $dir/lpd/brmfcl3770cdwfilter
     '';
 
     meta = {
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index f43c955befc..780fe3ff2ba 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -22895,6 +22895,7 @@ in
   mfcl2740dwcupswrapper = callPackage ../misc/cups/drivers/mfcl2740dwcupswrapper { };
   mfcl2740dwlpr = callPackage ../misc/cups/drivers/mfcl2740dwlpr { };
 
+  # This driver is only available as a 32 bit proprietary binary driver
   mfcl3770cdwlpr = (pkgsi686Linux.callPackage ../misc/cups/drivers/brother/mfcl3770cdw/default.nix { }).driver;
   mfcl3770cdwcupswrapper = (pkgsi686Linux.callPackage ../misc/cups/drivers/brother/mfcl3770cdw/default.nix { }).cupswrapper;
 
```

* * *

I'm wondering what is the difference for the end-user between this way (pkgsi686Linux.callPackage) and the way the other drivers do it (callPackage + using pkgsi686Linux.glibc). Seems like only printer drivers do the whole `interpreter=${pkgsi686Linux.glibc}/lib/ld-linux.so.2` thing.

Reply:
Who do you regard as the end-user? Looking at this after a while I would prefer to leave the details of the 32 bit inside the driver's expression

Reply:
Right, inside the expression is fine I think.

I'm thinking that hardcoding the location of ld-linux, while unlikely to change, might be an issue.

Considering the following:

```
nix-instantiate --eval -E \
    'with import <nixpkgs> {}; builtins.replaceStrings ["\n"] [""] (builtins.readFile "${pkgsi686Linux.stdenv.cc}/nix-support/dynamic-linker") == "${pkgsi686Linux.glibc}/lib/ld-linux.so.2"'
```

I don't know if there is a way to derive the `dynamic-linker` file from the i686 packages set instead of hardcoding the value. While unlikely to change, I'm not particularly fond of the hardcoded value, even though a few other derivations use the same principle. I guess that using `pkgsi686.stdenv.mkDerivation` would make `$NIX_CC` the right `cc` so the file would be available to cat as described. Though I do not either if it is bad for other reasons.

Reply:
I added the requested change so the dynamic-linker path is no longer hardcoded.